File: .cursor/rules/memory_anchors/example_anchors.js
/**
 * Example Memory Anchors Implementation - JavaScript
 *
 * This file demonstrates proper implementation of memory anchors
 * in a JavaScript codebase, showing various patterns and use cases.
 *
 * Maturity: beta
 *
 * Why:
 * - Provides concrete examples of memory anchor implementation in JavaScript
 * - Serves as a reference for front-end developers adding memory anchors
 * - Demonstrates best practices for anchor naming and categorization
 * - Shows how to use advanced memory anchor features in JS files
 */

// MEMORY_ANCHOR: {architectural} data_store_implementation
/**
 * DataStore - Centralized application data store
 * 
 * This class implements a reactive data store that:
 * 1. Provides a centralized state management solution
 * 2. Notifies subscribers of state changes
 * 3. Maintains state history for time-travel debugging
 * 4. Handles serialization and persistence
 */
class DataStore {
  constructor(options = {}) {
    // MEMORY_ANCHOR: {config} data_store_configuration
    // COMPLEXITY: medium
    // LAST_REVIEWED: 2023-11-10
    // REVIEWERS: @sarah, @miguel
    this.state = options.initialState || {};
    this.history = [];
    this.subscribers = [];
    this.maxHistoryLength = options.maxHistoryLength || 50;
    this.persistenceKey = options.persistenceKey;
    
    // Load persisted state if available
    if (this.persistenceKey && typeof window !== 'undefined') {
      try {
        const persisted = localStorage.getItem(this.persistenceKey);
        if (persisted) {
          this.state = JSON.parse(persisted);
          console.log(`Loaded persisted state from ${this.persistenceKey}`);
        }
      } catch (err) {
        console.error('Failed to load persisted state:', err);
      }
    }
  }
  
  // MEMORY_ANCHOR: {core} state_update_flow
  /**
   * Updates the store state
   * 
   * @param {Function|Object} updater - Function or object to update state
   * @param {Object} options - Update options
   * @returns {Object} New state after update
   */
  update(updater, options = {}) {
    const prevState = { ...this.state };
    
    // Handle function or object updater
    if (typeof updater === 'function') {
      this.state = updater(prevState);
    } else {
      this.state = { ...prevState, ...updater };
    }
    
    // Save to history
    if (!options.skipHistory) {
      this.history.push(prevState);
      
      // Prune history if needed
      if (this.history.length > this.maxHistoryLength) {
        this.history.shift();
      }
    }
    
    // Notify subscribers
    if (!options.silent) {
      this._notifySubscribers(prevState);
    }
    
    // Handle persistence
    if (this.persistenceKey && !options.skipPersistence) {
      this._persistState();
    }
    
    return this.state;
  }
  
  // MEMORY_ANCHOR: {api} subscription_management
  // RELATED_TO: state_update_flow
  /**
   * Subscribe to state changes
   * 
   * @param {Function} subscriber - Callback function to be called on state changes
   * @returns {Function} Unsubscribe function
   */
  subscribe(subscriber) {
    if (typeof subscriber !== 'function') {
      throw new Error('Subscriber must be a function');
    }
    
    this.subscribers.push(subscriber);
    
    // Return unsubscribe function
    return () => {
      const index = this.subscribers.indexOf(subscriber);
      if (index !== -1) {
        this.subscribers.splice(index, 1);
      }
    };
  }
  
  /**
   * Notify all subscribers of state change
   * 
   * @private
   * @param {Object} prevState - Previous state before update
   */
  _notifySubscribers(prevState) {
    for (const subscriber of this.subscribers) {
      try {
        subscriber(this.state, prevState);
      } catch (err) {
        console.error('Error in subscriber:', err);
      }
    }
  }
  
  /**
   * Persist state to localStorage
   * 
   * @private
   */
  _persistState() {
    if (typeof window !== 'undefined' && this.persistenceKey) {
      try {
        localStorage.setItem(this.persistenceKey, JSON.stringify(this.state));
      } catch (err) {
        console.error('Failed to persist state:', err);
      }
    }
  }
  
  // MEMORY_ANCHOR: {feature} time_travel_debugging
  /**
   * Travel back in state history
   * 
   * @param {number} steps - Number of steps to go back
   * @returns {Object} New state after time travel
   */
  timeTravel(steps = 1) {
    if (steps <= 0 || this.history.length === 0) {
      return this.state;
    }
    
    const actualSteps = Math.min(steps, this.history.length);
    const targetIndex = this.history.length - actualSteps;
    const newState = this.history[targetIndex];
    
    // Remove entries from history
    this.history = this.history.slice(0, targetIndex);
    
    // Update state without adding to history
    return this.update(() => newState, { skipHistory: true });
  }
  
  /**
   * Get a specific slice of the state
   * 
   * @param {string} path - Dot notation path to the state slice
   * @param {*} defaultValue - Default value if path doesn't exist
   * @returns {*} State slice or default value
   */
  get(path, defaultValue) {
    return this._getNestedProperty(this.state, path, defaultValue);
  }
  
  /**
   * Helper to retrieve nested property
   * 
   * @private
   * @param {Object} obj - Object to get property from
   * @param {string} path - Dot notation path
   * @param {*} defaultValue - Default value if path doesn't exist
   * @returns {*} Property value or default
   */
  _getNestedProperty(obj, path, defaultValue) {
    const keys = path.split('.');
    let result = obj;
    
    for (const key of keys) {
      if (result === undefined || result === null) {
        return defaultValue;
      }
      result = result[key];
    }
    
    return result !== undefined ? result : defaultValue;
  }
}

// MEMORY_ANCHOR: {utility} example_usage
/**
 * Example usage of the DataStore
 */
function exampleUsage() {
  // Create a store instance
  const store = new DataStore({
    initialState: {
      user: {
        name: 'John Doe',
        preferences: {
          theme: 'dark',
          notifications: true
        }
      },
      todos: []
    },
    persistenceKey: 'my-app-state'
  });
  
  // Subscribe to state changes
  const unsubscribe = store.subscribe((newState, prevState) => {
    console.log('State changed:', newState);
    renderApp(newState);
  });
  
  // Add a todo
  store.update(state => ({
    ...state,
    todos: [...state.todos, { id: 1, text: 'Learn about memory anchors', completed: false }]
  }));
  
  // Update user preferences
  store.update(state => ({
    ...state,
    user: {
      ...state.user,
      preferences: {
        ...state.user.preferences,
        theme: 'light'
      }
    }
  }));
  
  // Unsubscribe when done
  unsubscribe();
}

// Mock render function
function renderApp(state) {
  // DOM rendering logic would go here
}

// Export the store
module.exports = { DataStore }; 

File: .cursor/rules/memory_anchors/example_anchors.py
#!/usr/bin/env python3

"""Example Memory Anchors Implementation

This file demonstrates proper implementation of memory anchors
in a Python codebase, showing various patterns and use cases.

Maturity: beta

Why:
- Provides concrete examples of memory anchor implementation
- Serves as a reference for developers adding memory anchors
- Demonstrates best practices for anchor naming and categorization
- Shows how to use advanced memory anchor features
"""

import os
import json
import logging
from typing import Dict, List, Optional, Union, Any

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


# MEMORY_ANCHOR: {core} data_processing_pipeline
class DataProcessor:
    """Main data processing pipeline.
    
    This class implements the core data processing pipeline that handles:
    1. Data loading from various sources
    2. Data transformation and cleaning
    3. Feature extraction
    4. Output formatting
    """
    
    def __init__(self, config_path: str):
        # MEMORY_ANCHOR: {config} data_processor_initialization
        self.config = self._load_config(config_path)
        self.processors = []
        self.initialized = False
        logger.info("DataProcessor initialized with config from %s", config_path)
    
    def _load_config(self, config_path: str) -> Dict[str, Any]:
        """Load configuration from a JSON file."""
        try:
            with open(config_path, 'r') as f:
                return json.load(f)
        except Exception as e:
            logger.error("Failed to load config: %s", str(e))
            return {}
    
    # MEMORY_ANCHOR: {algorithmic} data_loading_algorithm
    # COMPLEXITY: medium
    # LAST_REVIEWED: 2023-10-15
    def load_data(self, source: str) -> List[Dict[str, Any]]:
        """Load data from the specified source.
        
        This function handles multiple data source types:
        - Local CSV/JSON files
        - Database connections
        - API endpoints
        - S3 buckets
        
        Returns a list of dictionaries representing the loaded data.
        """
        logger.info("Loading data from %s", source)
        
        if source.endswith('.json'):
            # JSON file loading logic
            with open(source, 'r') as f:
                return json.load(f)
        elif source.endswith('.csv'):
            # CSV file loading logic
            return self._load_csv(source)
        elif source.startswith('s3://'):
            # S3 bucket loading logic
            return self._load_from_s3(source)
        elif source.startswith('http'):
            # API endpoint loading logic
            return self._load_from_api(source)
        else:
            logger.error("Unsupported data source: %s", source)
            return []
    
    def _load_csv(self, filepath: str) -> List[Dict[str, Any]]:
        """Load data from a CSV file."""
        # Implementation details for CSV loading
        return []
    
    def _load_from_s3(self, s3_path: str) -> List[Dict[str, Any]]:
        """Load data from an S3 bucket."""
        # Implementation details for S3 loading
        return []
    
    def _load_from_api(self, api_url: str) -> List[Dict[str, Any]]:
        """Load data from an API endpoint."""
        # Implementation details for API loading
        return []
    
    # MEMORY_ANCHOR: {algorithmic} data_transformation
    # RELATED_TO: data_loading_algorithm, feature_extraction
    def transform_data(self, data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Transform and clean the input data.
        
        This function:
        1. Handles missing values
        2. Normalizes data formats
        3. Applies filters
        4. Performs data type conversions
        """
        logger.info("Transforming %d data records", len(data))
        
        transformed_data = []
        for record in data:
            # Apply transformations
            transformed_record = self._apply_transformations(record)
            if transformed_record:
                transformed_data.append(transformed_record)
        
        logger.info("Transformation complete. %d records after transformation", 
                   len(transformed_data))
        return transformed_data
    
    def _apply_transformations(self, record: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Apply transformations to a single record."""
        # Implementation details for transformations
        return record
    
    # MEMORY_ANCHOR: {algorithmic} feature_extraction
    def extract_features(self, data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Extract features from the transformed data.
        
        This function handles:
        1. Text feature extraction
        2. Numerical feature scaling
        3. Categorical encoding
        4. Feature normalization
        """
        logger.info("Extracting features from %d records", len(data))
        
        # Feature extraction implementation
        return data
    
    # MEMORY_ANCHOR: {core} process_pipeline_execution
    def process(self, source: str) -> List[Dict[str, Any]]:
        """Execute the full data processing pipeline.
        
        This is the main entry point for data processing.
        """
        logger.info("Starting data processing for source: %s", source)
        
        # Step 1: Load data
        data = self.load_data(source)
        if not data:
            logger.warning("No data loaded from %s", source)
            return []
        
        # Step 2: Transform data
        transformed_data = self.transform_data(data)
        if not transformed_data:
            logger.warning("No data after transformation")
            return []
        
        # Step 3: Extract features
        processed_data = self.extract_features(transformed_data)
        
        logger.info("Data processing complete. Processed %d records", len(processed_data))
        return processed_data


# MEMORY_ANCHOR: {utility} example_usage
def example_usage():
    """Demonstrate usage of the DataProcessor class."""
    # Create a processor instance
    processor = DataProcessor("config.json")
    
    # Process data from different sources
    json_data = processor.process("data.json")
    csv_data = processor.process("data.csv")
    api_data = processor.process("https://api.example.com/data")
    
    # Combine results
    all_data = json_data + csv_data + api_data
    
    # Write results to output file
    with open("processed_data.json", "w") as f:
        json.dump(all_data, f, indent=2)


if __name__ == "__main__":
    example_usage() 